torch/csrc/autograd/custom_function.h

```diff
@@ -18,33 +20,80 @@ TORCH_API variable_list _wrap_outputs(
+// many Variables as there were outputs from forward as arguments. It should
+//  return as many Variables as there were inputs with each of them containing
+// the gradient w.r.t. its corresponding input. Variables saved in forward can
+// be accessed with ctx->get_saved_variables() and other saved data can be
+// accessed from ctx->save.
 //
 // For example:
-// class MyFunction : public CFunction<MyFunction> {
+// class MyFunction : public Function<MyFunction> {
 //   public:
-//   static variable_list forward(AutogradContext *ctx, int n, Variable var);
+//   static variable_list forward(AutogradContext *ctx, int n, Variable var) {
+//      // Save data for backward in context
+//      ctx->save["n"] = n;
+//      var.mul_(2);
+//      // Mark var as modified by inplace operation
+//      ctx->mark_dirty({var});
+//      return std::vector<Variable>({var});
+//   }
 //
-//   static variable_list backward(AutogradContext *ctx, variable_list grad_output);
+//   static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+//      // Use data saved in forward
+//      auto n = ctx->save["n"].toInt();
+//      return std::vector<Variable>({grad_output[0]*n});
+//   }
 // };
+//
 // To use MyFunction
 // Variable x;
-// MyFunction::apply(6, x);
+// auto y = MyFunction::apply(6, x);
+// Example backward call:
+// y[0].sum().backward();
 template <class T>
 struct TORCH_API Function {
   template<typename... Args>
   static variable_list apply(Args&&... args);
 };
 
-struct AutogradContext {
-  std::vector<torch::autograd::SavedVariable> saved_variables;
-
+// Context to save information during forward that can be accessed in backward
+struct TORCH_API AutogradContext {
+  // Can be used to save non-variable data for backward()
+  ska::flat_hash_map<std::string, at::IValue> saved_data;
+
+  // Saves the list of variables for a future call to backward(). This
+  // should be called at most once from inside of forward().
+  void save_for_backward(const variable_list &to_save);
+  // Marks variables in the list as modified in an in-place opertaion. This
+  // should be called at most once from inside of forward() and all arguments
+  // should be inputs.
+  void mark_dirty(const variable_list &inputs);
+  // Marks outputs in the list as not requiring gradients. This should be called
+  // at most once from inside of forward() and all arguments should be outputs.
+  void mark_non_differentiable(const variable_list &outputs);
+  void clear_saved();
+
+  // Get the list of variables that were saved in forward using
+  // save_for_backward(). Before returning them to the user, a check is made to
+  // ensure that they were not modified by any in-place operations.
+  variable_list get_saved_variables() const;
+  const std::unordered_set<at::TensorImpl*>& get_dirty() const;
+  const std::unordered_set<at::TensorImpl*>& get_non_differentiable() const;
+
+private:
```

Comment:
OK good, I like the private specifier better. Actually, I wonder if `saved_data` shouldn't be private too. We could probably give a more user friendly way of setting and getting.